### DISCUSSION TITLE
XIPKI in docker container

### DISCUSSION THREAD

**Discussion by xipki** [xipki/xipki#249] [Show and tell]
As discussed in the issue #205, 
[fuchsia74](https://github.com/fuchsia74) created a git repository https://github.com/fuchsia74/xipki_dockerfiles containing the docker files. CA is working, OCSP not yet.

[plech-sec](https://github.com/plech-sec) created a docker image for OCSP with a CRL as the datasource:

( I used alpine instead of a tomcat docker image) 

A few important passages, below  :

```
FROM alpine:latest
MAINTAINER Pawel Lech 
# set env
ENV JAVA_HOME="/usr/lib/jvm/default-jvm"
ENV PATH="$JAVA_HOME/bin:$PATH"
ENV CATALINA_HOME="/opt/apache-tomcat-9.0.44"
```

```
RUN apk --no-cache add openjdk11 --repository=https://dl-cdn.alpinelinux.org/alpine/edge/community \
&& apk --no-cache add wget \
&& apk --no-cache add p7zip \
&& apk --no-cache add ca-certificates \
&& rm -rf /var/cache/apk/*

```

```
# create non-root user
RUN addgroup -S -g 10000 xipki \
&& adduser -S -u 10000 xipki -G xipki 

```

```
# fix permissions (especially for running as non-root)
# https://github.com/docker-library/tomcat/issues/35

RUN chmod -R 777 ${CATALINA_HOME}/logs \
&& chmod -R 777 ${CATALINA_HOME}/work \
&& chmod -R 777 ${CATALINA_HOME}/temp
```

```
WORKDIR ${CATALINA_HOME}
EXPOSE 8080
USER xipki
CMD ["/opt/apache-tomcat-9.0.44/bin/catalina.sh","run"]

```
The tomcat with **startup.sh** will start tomcat in the background (as a service) and docker will exit with code 0 which is normal and expected behavior.

In this case we need to start tomcat as foreground process with **catalina.sh run**

Sample OCSP response :

![response](https://user-images.githubusercontent.com/53973586/118379174-d9688080-b5d8-11eb-8f59-d8f488a53477.PNG)

**Comment by changhr2013** [comment]
I haven't seen a solution out of the box yet, but I'm trying to solve it using docker-compose, it is useful when you want to quickly verify xipki functionality。

**Reply by xipki** [reply]
Thanks. Any update?

**Reply by changhr2013** [reply]
everything is working fine, but I need to add some documentation for it.

**Reply by ralienpp** [reply]
Hi, do you have any news on this front? Perhaps you could share some intermediate work that we could build upon, and eventually make it a part of the official XiPKI repo?

**Reply by changhr2013** [reply]
> Hi, do you have any news on this front? Perhaps you could share some intermediate work that we could build upon, and eventually make it a part of the official XiPKI repo?

https://github.com/changhr2013/docker-xipki

I must admit that I only conducted very early-stage validation work. At the time, the code in this repository was able to pass validation successfully. However, I haven’t had much time to follow up and improve it since then, nor have I organized any practical documentation. I apologize for this. If I have some free time, I might revisit and improve it. Containerization can indeed significantly lower the deployment threshold.
